units/systemd-udev-trigger.service

```diff
@@ -20,4 +20,4 @@ ConditionPathIsReadWrite=/sys
 Type=oneshot
 RemainAfterExit=yes
 ExecStart=udevadm trigger --type=subsystems --action=add
-ExecStart=udevadm trigger --type=devices --action=add
+ExecStart=udevadm trigger --type=devices --subsystem-nomatch=memory --action=add
```

Comment:
Hmm, I always thought that this coldplug phase is necessary to catch hotplug events that came in between the start of the system and the time when udev became active.

Reply:
> Hmm, I always thought that this coldplug phase is necessary to catch hotplug events that came in between the start of the system and the time when udev became active.

So did I. But our kernel mm experts have assured me that no user-space initialization is necessary for these devices. The events are really only needed for handling **real** hot-plug events. Note that upstream systemd has no single rule for these devices. udev doesn't create db files for them (because no non-trivial environment variables are set for them, I suppose). The way `udevadm trigger`  works, it writes to 10000s of `uevent` files in sysfs, which causes kernel uevents, which spawns lots of workers accessing the same sysfs subhierarchy, while `trigger` is still writing. Together, this causes massive contention.

The most important consumers of this functionality are very large systems with support for physical or logical partitioning of the hardware, where memory can be moved between partitions. And it's this kind of systems for which the HW designers are explicitly asking us turn these coldplug events off. It's particularly bad on PowerPC because the small granularity of the memory devices (256 MB, which is really not much on systems that can have several TB of physical RAM).

I would have preferred a different solution as well, but over extensive discussions with hardware and kernel people, this has turned out to be the only realistic approach. It has been suggested to decrease the number of memory devices by increasing the granularity, which is fair enough, but the currently existing hardware is just the way it is and can't be changed.

AFAIU, at the heart of the problem, there are inefficient implementations locks on the PowerPC architecture in general, and inefficient handling of high numbers of concurrent accesses in kernfs in particular. The kernfs maintainer has basically shrugged this off (see https://lkml.org/lkml/2020/6/25/192 and related posts if you're intrested in the details).

I proposed this on SUSE's downstream repo first (https://github.com/openSUSE/systemd/pull/23). @fbuihuu asked me to discuss it here.